core/lib/src/local/response.rs

```diff
@@ -180,6 +180,134 @@ macro_rules! pub_response_impl {
+    /// implement `RouteType`, see [`RouteType`](crate::route::RouteType) for more information.
+    pub fn was_routed_by<T: crate::route::RouteType>(&self) -> bool {
+        // If this request was caught, the route in `.route()` did NOT generate this response.
+        if self._request().catcher().is_some() {
+            false
+        } else if let Some(route_type) = self._request().route().map(|r| r.route_type).flatten() {
+            route_type == std::any::TypeId::of::<T>()
+        } else {
+            false
+        }
+    }
+
+    /// Checks if a request was routed to a specific route type. This will return true for routes
+    /// that were attempted, _but not actually called_. This enables a test to verify that a route
+    /// was attempted, even if another route actually generated the response, e.g. an
+    /// authenticated route will typically defer to an error catcher if the request does not have
+    /// the proper authentication. This makes it possible to verify that a request was routed to
+    /// the authentication route, even if the response was eventaully generated by another route or
+    /// a catcher.
+    ///
+    /// # Example
+    ///
+    // WARNING: this doc-test is NOT run, because cargo test --doc does not run doc-tests for items
+    // only available during tests.
+    /// ```rust
+    /// # use rocket::{get, routes, async_trait, request::{Request, Outcome, FromRequest}};
+    /// # struct WillFail {}
+    /// # #[async_trait]
+    /// # impl<'r> FromRequest<'r> for WillFail {
+    /// #     type Error = ();
+    /// #     async fn from_request(request: &'r Request<'_>) -> Outcome<Self, Self::Error> {
+    /// #         Outcome::Forward(())
+    /// #     }
+    /// # }
+    /// #[get("/", rank = 2)]
+    /// fn index1(guard: WillFail) -> &'static str { "Hello World" }
+    /// #[get("/")]
+    /// fn index2() -> &'static str { "Hello World" }
+    #[doc = $doc_prelude]
+    /// # Client::_test_with(|r| r.mount("/", routes![index1, index2]), |_, _, response| {
+    /// let response: LocalResponse = response;
+    /// assert!(response.was_attempted_by::<index1>());
+    /// assert!(response.was_attempted_by::<index2>());
+    /// assert!(response.was_routed_by::<index2>());
+    /// # });
+    /// ```
+    ///
+    /// # Other Route types
+    ///
+    /// [`FileServer`](crate::fs::FileServer) implementes `RouteType`, so a route that should
+    /// return a static file can be checked against it. Libraries which provide custom Routes should
+    /// implement `RouteType`, see [`RouteType`](crate::route::RouteType) for more information.
+    ///
+    /// # Note
+    ///
+    /// This method is marked as `cfg(test)`, and is therefore only available in unit and
+    /// integration tests. This is because the list of routes attempted is only collected in these
+    /// testing environments, to minimize performance impacts during normal operation.
+    #[cfg(test)]
```

Comment:
This is unfortunately not the way `cfg(test)` works. `cfg(test)` only gets enabled on the top-level crate being tested, not on its dependencies. So `rocket` compiled as a dependency of an application, even when compiling that application's tests, will never see this `cfg` enabled.